_aicspylibczi/ImagesContainer.h

```diff
@@ -0,0 +1,103 @@
+//
+// Created by Jamie Sherman on 7/22/20.
+//
+
+#ifndef _AICSPYLIBCZI_IMAGESCONTAINER_H
+#define _AICSPYLIBCZI_IMAGESCONTAINER_H
+
+#include <memory>
+
+#include "Image.h"
+
+namespace pylibczi {
+
+  template<typename T> class ImagesContainer;
+
+  class ImagesContainerBase {
+  public:
+      using Shape = std::vector<std::pair<char, size_t> >;
+      using ImagesContainerBasePtr = std::unique_ptr<ImagesContainerBase>;
+
+  private:
+      ImageVector m_images;
+      Shape m_shape;
+      libCZI::PixelType m_cziPixelType;
+
+  public:
+      static ImagesContainerBasePtr getTypedAsBase(libCZI::PixelType& pixel_type_, size_t pixels_in_all_images_);
+
+      template<typename T>
+      ImagesContainer<T>* getBaseAsTyped(void)
+      {
+          return static_cast< ImagesContainer<T>*>(this);
```

Comment:
this should probably be dynamic_cast since it is actually a derived class... in fact why not just call dynamic_cast instead of getBaseAsTyped?

Reply:
Should I be using dynamic_cast or dynamic_pointer_cast? I'm reading the docs and trying to figure it out but it looks like there's a subtlety I think I'm missing.

Reply:
It looked to me like `getBaseAsTyped` could be completely replaced with `dynamic_cast`.  Or if you feel more comfortable, replace its internal static_cast with dynamic_cast.   `dynamic_pointer_cast` is relevant for `shared_ptr`s only and is not pertinent for this code.

Reply:
So there's a template/polymorphism clash happening here so I think I'm stuck with static_cast.

Reply:
The trick to getting this going was as @toloudis suggested, adding a virtual destructor to the base class. Then polymorphism with the templated child class works with dynamic_cast.